File: src/main/java/marisa/powers/Marisa/SuperNovaPower.java
package marisa.powers.Marisa;

import marisa.marisa;
import com.badlogic.gdx.graphics.Texture;
import com.megacrit.cardcrawl.actions.common.ApplyPowerAction;
import com.megacrit.cardcrawl.actions.common.ExhaustSpecificCardAction;
import com.megacrit.cardcrawl.cards.AbstractCard;
import com.megacrit.cardcrawl.cards.AbstractCard.CardType;
import com.megacrit.cardcrawl.cards.status.Burn;
import com.megacrit.cardcrawl.characters.AbstractPlayer;
import com.megacrit.cardcrawl.core.AbstractCreature;
import com.megacrit.cardcrawl.core.CardCrawlGame;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.localization.PowerStrings;
import com.megacrit.cardcrawl.powers.AbstractPower;
import com.megacrit.cardcrawl.powers.StrengthPower;

public class SuperNovaPower extends AbstractPower {

  public static final String POWER_ID = "SuperNovaPower";
  private static final PowerStrings powerStrings = CardCrawlGame.languagePack
      .getPowerStrings(POWER_ID);
  public static final String NAME = powerStrings.NAME;
  public static final String[] DESCRIPTIONS = powerStrings.DESCRIPTIONS;
  private AbstractPlayer p;

  public SuperNovaPower(AbstractCreature owner, int amount) {
    this.name = NAME;
    this.ID = POWER_ID;
    this.owner = owner;
    this.amount = amount;
    this.type = AbstractPower.PowerType.BUFF;
    updateDescription();
    this.img = new Texture("img/powers/impulse.png");
    this.p = AbstractDungeon.player;
  }

  public void atEndOfTurn(boolean isPlayer) {
    if (!this.p.hand.isEmpty()) {
      flash();
      for (AbstractCard c : this.p.hand.group) {
        AbstractDungeon.actionManager.addToBottom(
            new ExhaustSpecificCardAction(c, this.p.hand));
      }
    }
  }

  public void onExhaust(AbstractCard card) {
    boolean apply = (card instanceof Burn);
    if (apply) {
      AbstractDungeon.actionManager.addToBottom(
          new ApplyPowerAction(p, p, new StrengthPower(p, this.amount), this.amount));
    }
  }

  @Override
  public void onDrawOrDiscard() {
    //ThMod.logger.info("SuperNovaPower : onDrawOrDiscard : ExhaustDiscard");
    ExhaustDiscard();
  }

  @Override
  public void onApplyPower(AbstractPower power, AbstractCreature target, AbstractCreature source) {
    //ThMod.logger.info("SuperNovaPower : onApplyPower : ExhaustDiscard");
    ExhaustDiscard();
  }

  @Override
  public void onInitialApplication() {
    //ThMod.logger.info("SuperNovaPower : onInitialApplication : ExhaustDiscard");
    ExhaustDiscard();
  }

  private void ExhaustDiscard() {
    for (AbstractCard c : AbstractDungeon.player.hand.group) {
      if (discardCheck(c)) {
        c.exhaust = true;
        c.isEthereal = true;
      }
    }
  }

  private boolean discardCheck(AbstractCard card) {
    if ((card.type == CardType.CURSE) || (card.type == CardType.STATUS)) {
      marisa.logger.info("SuperNovaPower : discardCheck : " + card.cardID + " detected.");
      return true;
    }
    return false;
  }

  public void updateDescription() {
      this.description = (DESCRIPTIONS[0] + this.amount + DESCRIPTIONS[2]);
  }

}

File: src/main/kotlin/marisa/powers/Marisa/SuperNovaPower.kt
package marisa.powers.Marisa

import com.badlogic.gdx.graphics.Texture
import com.megacrit.cardcrawl.actions.common.ApplyPowerAction
import com.megacrit.cardcrawl.actions.common.ExhaustSpecificCardAction
import com.megacrit.cardcrawl.cards.AbstractCard
import com.megacrit.cardcrawl.cards.AbstractCard.CardType
import com.megacrit.cardcrawl.cards.status.Burn
import com.megacrit.cardcrawl.characters.AbstractPlayer
import com.megacrit.cardcrawl.core.AbstractCreature
import com.megacrit.cardcrawl.core.CardCrawlGame
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.powers.AbstractPower
import com.megacrit.cardcrawl.powers.StrengthPower
import marisa.MarisaContinued

class SuperNovaPower(owner: AbstractCreature?, amount: Int) : AbstractPower() {
    private val p: AbstractPlayer

    init {
        name = NAME
        ID = POWER_ID
        this.owner = owner
        this.amount = amount
        type = PowerType.BUFF
        updateDescription()
        img = Texture("img/powers/impulse.png")
        p = AbstractDungeon.player
    }

    override fun atEndOfTurn(isPlayer: Boolean) {
        if (!p.hand.isEmpty) {
            flash()
            for (c in p.hand.group) {
                addToBot(
                    ExhaustSpecificCardAction(c, p.hand)
                )
            }
        }
    }

    override fun onExhaust(card: AbstractCard) {
        val apply = card is Burn
        if (apply) {
            addToBot(
                ApplyPowerAction(p, p, StrengthPower(p, amount), amount)
            )
        }
    }

    override fun onDrawOrDiscard() {
        //ThMod.logger.info("SuperNovaPower : onDrawOrDiscard : ExhaustDiscard");
        ExhaustDiscard()
    }

    override fun onApplyPower(
        power: AbstractPower,
        target: AbstractCreature,
        source: AbstractCreature
    ) {
        //ThMod.logger.info("SuperNovaPower : onApplyPower : ExhaustDiscard");
        ExhaustDiscard()
    }

    override fun onInitialApplication() {
        //ThMod.logger.info("SuperNovaPower : onInitialApplication : ExhaustDiscard");
        ExhaustDiscard()
    }

    private fun ExhaustDiscard() {
        for (c in AbstractDungeon.player.hand.group) {
            if (discardCheck(c)) {
                c.exhaust = true
                c.isEthereal = true
            }
        }
    }

    private fun discardCheck(card: AbstractCard): Boolean {
        if (card.type == CardType.CURSE || card.type == CardType.STATUS) {
            MarisaContinued.logger.info("SuperNovaPower : discardCheck : " + card.cardID + " detected.")
            return true
        }
        return false
    }

    override fun updateDescription() {
        description = DESCRIPTIONS[0] + amount + DESCRIPTIONS[2]
    }

    companion object {
        const val POWER_ID = "SuperNovaPower"
        private val powerStrings = CardCrawlGame.languagePack
            .getPowerStrings(POWER_ID)
        val NAME = powerStrings.NAME
        val DESCRIPTIONS = powerStrings.DESCRIPTIONS
    }
}
